src/java/org/apache/cassandra/db/virtual/AccordDebugKeyspace.java

```diff
@@ -0,0 +1,671 @@
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.marshal.ByteBufferAccessor;
+import org.apache.cassandra.db.marshal.Int32Type;
+import org.apache.cassandra.db.marshal.LongType;
+import org.apache.cassandra.db.marshal.TupleType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UUIDType;
+import org.apache.cassandra.dht.Range;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.exceptions.InvalidRequestException;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.accord.AccordCommandStore;
+import org.apache.cassandra.service.accord.AccordKeyspace;
+import org.apache.cassandra.service.accord.AccordService;
+import org.apache.cassandra.service.accord.AccordStateCache;
+import org.apache.cassandra.service.accord.CommandStoreTxnBlockedGraph;
+import org.apache.cassandra.service.accord.api.AccordRoutingKey;
+import org.apache.cassandra.service.accord.api.AccordRoutingKey.TokenKey;
+import org.apache.cassandra.service.consensus.migration.ConsensusMigrationState;
+import org.apache.cassandra.service.consensus.migration.TableMigrationState;
+import org.apache.cassandra.tcm.ClusterMetadata;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.Pair;
+
+import static java.lang.String.format;
+import static java.util.Comparator.comparing;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static accord.utils.async.AsyncChains.getBlockingAndRethrow;
+import static org.apache.cassandra.schema.SchemaConstants.VIRTUAL_ACCORD_DEBUG;
+import static org.apache.cassandra.utils.ByteBufferUtil.bytes;
+import static org.apache.cassandra.utils.MonotonicClock.Global.approxTime;
+
+public class AccordDebugKeyspace extends VirtualKeyspace
+{
+    // {epoch, hlc, flags, node}
+    private static final TupleType TIMESTAMP_TYPE =
+        new TupleType(List.of(LongType.instance, LongType.instance, Int32Type.instance, Int32Type.instance));
+    private static final String TIMESTAMP_TYPE_STRING = TIMESTAMP_TYPE.asCQL3Type().toString();
+
+    // {epoch, hlc, kind, domain, node}
+    private static final TupleType TXN_ID_TYPE =
+        new TupleType(List.of(LongType.instance, LongType.instance, UTF8Type.instance, UTF8Type.instance, Int32Type.instance));
+    private static final String TXN_ID_TYPE_STRING = TXN_ID_TYPE.asCQL3Type().toString();
+
+    // {table_id, token} or {table_id, +Inf/-Inf}
+    private static final TupleType ROUTING_KEY_TYPE = new TupleType(List.of(UUIDType.instance, UTF8Type.instance));
+    private static final String ROUTING_KEY_TYPE_STRING = ROUTING_KEY_TYPE.asCQL3Type().toString();
+
+    public static final AccordDebugKeyspace instance = new AccordDebugKeyspace();
+
+    private AccordDebugKeyspace()
+    {
+        super(VIRTUAL_ACCORD_DEBUG, List.of(
```

Comment:
why is `CoordinationStatus` dropped?  I use this for debugging deployed clusters...

Reply:
How was it of value? The initial coordination should last at most 10s (until it times out), after which all the relevant state for debugging in progress transactions should be found in the progress log table.

We could perhaps enrich the progress log home state to include information about active coordinations, including more useful information such as the phase we have reached. This would provide a more consistent means of monitoring this.

Reply:
> How was it of value?

It had the same value as the other nodetool commands we use to see what queries are actively running.  When I see user latencies start to regress or fall off a wall I have been using it to see what is going on and what is in-flight.

> We could perhaps enrich the progress log home state to include information about active coordinations, including more useful information such as the phase we have reached. This would provide a more consistent means of monitoring this.

I think we need a table that tells operators what is going on, and isn't expected to break in any random release... I agree this table has the 10s weakness you point out, but we need something.  If the progress log tables are expected to break w/e then we don't have the visibility we need.

Reply:
> I think we need a table that tells operators what is going on, and isn't expected to break in any random release... I agree this table has the 10s weakness you point out, but we need something. If the progress log tables are expected to break w/e then we don't have the visibility we need.

I think we can and should expose something like that before final release, but maybe delay until we are closer to completion and things are more-or-less fully settled? I've kept a skeleton `AccordVirtualTables` class to put in whatever tables we want to expose to the wider audience as final and stable later.

Reply:
>  but maybe delay until we are closer to completion and things are more-or-less fully settled?

I am fine with that.  If we want to say we want a coordination table but we are not yet ready to implement one, then I am fine removing in this patch.

Reply:
I’ll try to add that tonight

Reply:
It is a bit more complicated than I had expected to do a decent job of this, so I'm going to add a ticket to track adding this in future.